internal/controller/controllers/clusterdeployments_controller.go

```diff
@@ -426,6 +443,91 @@ func (r *ClusterDeploymentsReconciler) updateInstallConfigOverrides(ctx context.
+			_ = r.Manifests.DeleteClusterManifestInternal(ctx, operations.DeleteClusterManifestParams{
+				ClusterID: *cluster.ID,
+				FileName:  manifest.FileName,
+				Folder:    swag.String(models.ManifestFolderOpenshift),
+			})
+		}
+	}
+
+	// create/update all manifests provided by configmap data
+	for filename, manifest := range manifestsFromConfigMap {
+		r.Log.Infof("Creating cluster deployment %s manifest %s", cluster.KubeKeyName, filename)
+		_, err := r.Manifests.CreateClusterManifestInternal(ctx, operations.CreateClusterManifestParams{
+			ClusterID: *cluster.ID,
+			CreateManifestParams: &models.CreateManifestParams{
+				Content:  swag.String(base64.StdEncoding.EncodeToString([]byte(manifest))),
+				FileName: swag.String(filename),
+				Folder:   swag.String(models.ManifestFolderOpenshift),
+			}})
+		if err != nil {
+			r.Log.WithError(err).Errorf("Failed to create cluster deployment %s manifest %s", cluster.KubeKeyName, filename)
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *ClusterDeploymentsReconciler) getClusterDeploymentManifest(ctx context.Context, clusterDeployment *hivev1.ClusterDeployment) (map[string]string, error) {
+	configuredManifests := &corev1.ConfigMap{}
+	configuredManifests.Data = map[string]string{}
+	// get manifests from configmap if we have reference for it
+	if clusterDeployment.Spec.Provisioning.ManifestsConfigMapRef != nil {
+		err := r.Get(ctx, types.NamespacedName{Namespace: clusterDeployment.Namespace,
+			Name: clusterDeployment.Spec.Provisioning.ManifestsConfigMapRef.Name}, configuredManifests)
+		if err != nil {
+			r.Log.WithError(err).Errorf("Failed to get configmap %s in %s",
+				clusterDeployment.Spec.Provisioning.ManifestsConfigMapRef.Name, clusterDeployment.Namespace)
+			return nil, err
+		}
+	}
+	return configuredManifests.Data, nil
+}
+
+func (r *ClusterDeploymentsReconciler) addCustomManifests(ctx context.Context, clusterDeployment *hivev1.ClusterDeployment,
+	cluster *common.Cluster) error {
+
+	// added just in case we it will run after cluster will be in the middle of installation
+	if !common.IsInPreInstallationState(cluster) {
+		r.Log.Debugf("Cluster is not in pre installation state, skipping")
+		return nil
+	}
+
+	alreadyCreatedManifests, err := r.Manifests.ListClusterManifestsInternal(ctx, operations.ListClusterManifestsParams{
+		ClusterID: *cluster.ID,
+	})
+	if err != nil {
+		r.Log.WithError(err).Errorf("Failed to list manifests for %q cluster deployment", clusterDeployment.Name)
+		return err
+	}
+
+	// if reference to manifests was deleted from cluster deployment
+	// but we already added some in previous reconcile loop, we want to clean them.
+	// if no reference were provided we will delete all manifests that were in the list
+	if clusterDeployment.Spec.Provisioning.ManifestsConfigMapRef == nil && len(alreadyCreatedManifests) < 1 {
+		r.Log.Debugf("Nothing to do, skipping manifest creation")
```

Comment:
Are you expecting -3 `alreadyCreatedManifests`? why not check if `len(alreadyCreatedManifests) == 0`?